### PR TITLE
Refactor trap handler with linkme::distributed_slice

### PR DESCRIPTION
**Pull request by equation314**
It allows trap handlers to be defined in any crate.

### CODE REVIEW

File: modules/axhal/src/trap.rs [hunk 1]
```diff
@@ -1,23 +1,31 @@
...
-///
-/// [1]: crate_interface::def_interface
-/// [2]: crate_interface::impl_interface
-#[def_interface]
-pub trait TrapHandler {
-    /// Handles interrupt requests for the given IRQ number.
-    fn handle_irq(irq_num: usize);
-    // more e.g.: handle_page_fault();
-}
+pub use linkme::distributed_slice as reg_trap_handler;
```
Comment (hky1999) [inline modules/axhal/src/trap.rs]: what does 'reg' in `reg_trap_handler` mean?

File: scripts/make/cargo.mk [hunk 2]
```diff
@@ -17,7 +17,7 @@ build_args := \
   $(build_args-$(MODE)) \
   $(verbose)
 
-RUSTFLAGS := -C link-arg=-T$(LD_SCRIPT) -C link-arg=-no-pie
+RUSTFLAGS := -C link-arg=-T$(LD_SCRIPT) -C link-arg=-no-pie -C link-arg=-znostart-stop-gc
```
Comment (hky1999) [inline scripts/make/cargo.mk:20]: I this place need an annotation, like add refs about what is 'nostart-stop-gc', I find something here: https://releases.llvm.org/14.0.0/tools/lld/docs/ELF/start-stop-gc.html

File: ulib/axstd/src/os.rs [hunk 3]
```diff
@@ -3,4 +3,6 @@
 /// ArceOS-specific definitions.
 pub mod arceos {
     pub use arceos_api as api;
+    #[doc(no_inline)]
+    pub use arceos_api::modules;
```
Comment (hky1999) [inline ulib/axstd/src/os.rs:7]: Previously, ning did not expose these modules. What was ning's consideration for exposing all of them in the API now? Is there any safety concern if methods or vars in those modules is not exported as `pub(crate)`

File: modules/axhal/src/irq.rs [hunk 4]
```diff
@@ -33,3 +34,11 @@ pub(crate) fn register_handler_common(irq_num: usize, handler: IrqHandler) -> bo
     warn!("register handler for IRQ {} failed", irq_num);
     false
 }
+
+#[reg_trap_handler(IRQ)]
```
Comment (hky1999) [inline modules/axhal/src/irq.rs]: no `#[reg_trap_handler(PAGE_FAULT)]` currently ,right?

**Review by hky1999 — CHANGES_REQUESTED** [review]

File: modules/axhal/src/trap.rs (hunk 1 above)
Comment (equation314) [reply modules/axhal/src/trap.rs]: register

File: scripts/make/cargo.mk (hunk 2 above)
Comment (equation314) [reply scripts/make/cargo.mk:20]: https://github.com/dtolnay/linkme/issues/94

File: ulib/axstd/src/os.rs (hunk 3 above)
Comment (equation314) [reply ulib/axstd/src/os.rs:7]: To use `axhal` in apps without adding it to dependencies.

File: modules/axhal/src/irq.rs (hunk 4 above)
Comment (equation314) [reply modules/axhal/src/irq.rs]: Not required for the unikernels.

**Review by hky1999 — APPROVED** [review]

**Pull request merged**